src/iuse.cpp

```diff
@@ -8083,6 +8085,324 @@ std::optional<int> iuse::capture_monster_act( Character *p, item *it, const trip
+            heating_effect = it->type->charges_to_use();
+            p->add_msg_if_player( m_info, _( "You use %1$s to start heating." ), heater->tname() );
+        } else if( !it->has_no_links() ) {
+            heater = item_location( *p, it );
+            available_heater = it->link().t_veh->connected_battery_power_level().first;
+            heating_effect = it->type->charges_to_use();
+            p->add_msg_if_player( m_info, _( "You use %1$s to start heating." ), heater->tname() );
+        } else if( it->has_flag( flag_USE_UPS ) &&
+                   units::to_kilojoule( p->available_ups() ) >= it->type->charges_to_use() ) {
+            heater = item_location( *p, it );
+            available_heater = units::to_kilojoule( p->available_ups() );
+            heating_effect = it->type->charges_to_use();
+        } else {
+            // You can't put hotplate on other hotplate to heat items.
+            p->add_msg_if_player( m_info, _( "The %s's batteries are dead." ), it->tname() );
+            return std::nullopt;
+        }
+    } else if( !it->has_quality( qual_HOTPLATE ) ) {
+        auto filter = [p]( const item & e ) {
+            if( e.has_quality( qual_HOTPLATE, 2 ) && e.ammo_remaining() >= e.type->charges_to_use() ) {
+                return true;
+            }
+            if( e.has_quality( qual_HOTPLATE, 2 ) && ( !e.has_no_links() ) ) {
+                if( e.link().t_veh->connected_battery_power_level().first >= e.type->charges_to_use() ) {
+                    return true;
+                }
+            }
+            if( e.has_quality( qual_HOTPLATE, 2 ) && e.has_flag( flag_USE_UPS ) ) {
+                if( units::to_kilojoule( p->available_ups() ) >= e.type->charges_to_use() ) {
+                    return true;
+                }
+            }
+            return false;
+        };
+        heater = g->inv_map_splice( filter, _( "Select a tool to heat:" ), 1,
+                                    _( "You don't have proper heating source." ) );
+        if( !heater ) {
+            add_msg( m_info, _( "Never mind." ) );
+            return std::nullopt;
+        }
+        p->add_msg_if_player( m_info, _( "You put %1$s on %2$s to start heating." ), it->tname(),
+                              heater->tname() );
+        if( !heater->has_no_links() ) {
+            available_heater = heater->link().t_veh->connected_battery_power_level().first;
+            heating_effect = heater->type->charges_to_use();
+        } else if( !heater->has_flag( flag_USE_UPS ) ) {
+            available_heater = heater->ammo_remaining();
+            heating_effect = heater->type->charges_to_use();
+        } else if( heater->has_flag( flag_USE_UPS ) ) {
+            add_msg( m_info, _( "%1$i" ), units::to_kilojoule( p->available_ups() ) );
+            available_heater = units::to_kilojoule( p->available_ups() );
+            heating_effect = heater->type->charges_to_use();
+        }
+    }
```

Comment:
This whole section could (and should) be moved to a separate helper function, `find_heater` (or similar).
You can return a struct with the information you're populating here.